src/Microsoft.ML.Transforms/WhiteningTransform.cs

```diff
@@ -383,249 +507,339 @@ private void TrainModels(Float[][] columnData, int[] rowCounts, IChannel ch)
-
-            _models = new Float[Infos.Length][];
-            InvModels = new Float[Infos.Length][];
-            for (int i = 0; i < Infos.Length; i++)
-            {
-                _models[i] = ctx.Reader.ReadFloatArray();
-                ValidateModel(Host, _models[i], Infos[i].TypeSrc);
-                if (_exes[i].SaveInv)
-                {
-                    InvModels[i] = ctx.Reader.ReadFloatArray();
-                    ValidateModel(Host, InvModels[i], Infos[i].TypeSrc);
-                }
-            }
-            Metadata.Seal();
-        }
-
-        public static WhiteningTransform Create(IHostEnvironment env, ModelLoadContext ctx, IDataView input)
-        {
-            Contracts.CheckValue(env, nameof(env));
-            var h = env.Register(RegistrationName);
-            h.CheckValue(ctx, nameof(ctx));
-            h.CheckValue(input, nameof(input));
-            ctx.CheckAtModel(GetVersionInfo());
-
-            // *** Binary format ***
-            // int: sizeof(Float)
-            // <remainder handled in ctors>
-            int cbFloat = ctx.Reader.ReadInt32();
-            h.CheckDecode(cbFloat == sizeof(Float));
-            return h.Apply("Loading Model", ch => new WhiteningTransform(h, ctx, input));
-        }
-
         public override void Save(ModelSaveContext ctx)
         {
             Host.CheckValue(ctx, nameof(ctx));
             ctx.CheckAtModel();
             ctx.SetVersionInfo(GetVersionInfo());
 
             // *** Binary format ***
-            // int: sizeof(Float)
+            // int: sizeof(float)
             // <base>
             // foreach added column
             //   ColInfoEx
             // foreach model
             //   whitening matrix
             //   recovery matrix
-            ctx.Writer.Write(sizeof(Float));
 
-            SaveBase(ctx);
+            SaveColumns(ctx);
 
-            Host.Assert(_exes.Length == Infos.Length);
-            for (int i = 0; i < _exes.Length; i++)
-                _exes[i].Save(ctx);
+            Host.Assert(_infos.Length == ColumnPairs.Length);
+            for (int i = 0; i < _infos.Length; i++)
+                _infos[i].Save(ctx);
             for (int i = 0; i < _models.Length; i++)
             {
                 ctx.Writer.WriteFloatArray(_models[i]);
-                if (_exes[i].SaveInv)
-                    ctx.Writer.WriteFloatArray(InvModels[i]);
+                if (_infos[i].SaveInv)
+                    ctx.Writer.WriteFloatArray(_invModels[i]);
             }
         }
 
-        private static string TestColumn(ColumnType t)
+        private static class Mkl
         {
-            string reason = TestIsKnownSizeFloatVector(t);
-            if (reason != null)
-                return reason;
+            private const string DllName = "MklImports";
 
-            if ((long)t.ValueCount * t.ValueCount > Utils.ArrayMaxSize)
-                return "Vector size exceeds limit";
+            public enum Layout
```

Comment:
Could you please add a reference link to `Layout` and `Transpose`? It'd be very helpful because sometime search engine is not able to find those information.

Reply:
```suggestion
            // The allowed value of Layout is specified in Intel's MLK library. See Layout parameter in this [doc](https://software.intel.com/en-us/mkl-developer-reference-c-cblas-gemm) for details.
            public enum Layout
```

Reply:
My expectation is that, those enum's are not very intuitive to users and external devs, so we need to put some hints around them.